### ISSUE TITLE
[IMPORTANT] New version ModernMT v4

### ISSUE THREAD

**Issue by davidecaroselli** [modernmt/modernmt#449]
Hello all!

We have an important announcement to make, we are pleased to announce a brand new version or ModernMT software (v4.0) that will be available shortly and, with it, lot of new stuff!

The brand new version of ModernMT introduces some major innovations in our translation tool. ModernMT v4.0 will have:
- A brand new Neural Engine based on fairseq framework with **twice the translation speed**.
- A **boost in baseline models translation performance** measured in up to +2 BLEU points thanks to the new training pipeline.
- A brand **new aligner cleaning technique** to filter out noisy, un-aligned segments in training data for better performance.

Some minor changes to this repository:
- We will rename the repository from `modernmt/MMT` to `modernmt/modernmt`. So please, update your remote paths in git as soon as we update the name (I will write it here more details on how to do that).
- With the new name, the new code will replace the old `decoder-neural`.

Stay tuned for updates!

Edit: we decided to keep the project fully **open source**, this means that the code base **will not change Licensing anymore**, all new software will be released with Apache-2.0 as always!

**Comment by EtienneAb3d** [comment]
This means that MMT4 won't be any more free for commercial use ?

**Comment by davidecaroselli** [comment]
Hi @EtienneAb3d 

we are still validating with our legal dep. which is the best License to use, in fact, I'm editing the original post in order to remove misleading infos (CommonsClause must still be validated).

Our goal for the new software is to be free to use for research institutions and for any private use. For commercial use, the idea is to provide a commercial license so yes, ModernMT v4 won't be free for commercial use anymore.

However, regardless on which license we will end up with, we don't see any reason for not giving the free use for your NeuroSpell project!

**Comment by LoekvanKooten** [comment]
Is that one commercial license, like a 20.000 USD license for large enterprises, or have you also thought about one-man companies like mine?

**Comment by davidecaroselli** [comment]
@LoekvanKooten we have no interest into be a problem for you, we will create different options for sure. At the moment we have not figured out yet the commercial license and plans so, thanks for all your feedbacks, they will definitely drive our decisions.

**Comment by LoekvanKooten** [comment]
Thank you Davide, much appreciated!

**Comment by EtienneAb3d** [comment]
Yes, NeuroSpell is, for the moment, a dev on my free time. If it will get a success, I would certainly create a company with commercial purposes. I think of a freemium version: free for ordinary online uses, and payable for large needs and/or API access.
I'm here for a long time for my current employer: Lingua et Machina. We are using MMT for a commercial use. We are a small company. It's not possible for us to think paying for MMT. If MMT will become a payable software, certainly we will think to switch to some others.. or simply stay with MMT3, suppose this will force me to work more to code by myself to make it evolve, without having really the time for this, and without the benefit of your own future work. 
Will MMT3 license also be changed ?
More than this, I'm contributing here, at my level, with the idea that all my work/feedback/suggestions are available for a large public, including you and me. I don't think I would enjoy to contribute in the same way if this won't be the case any more. Perhaps only to ask for a new feature or a bug to be fixed.

**Comment by davidecaroselli** [comment]
@EtienneAb3d very quickly:

MMT3 **will not** change license (this is not even possible from a legally speaking) any change will impact only new versions of ModernMT

It is clear to us that ModernMT is currently used from very-small to very-large enterprises and our solution will be very aware of this difference.

**Comment by EtienneAb3d** [comment]
We are software developers. We are very small, but we are selling MMT solutions in our own product, and/or data processed by it, to any kind of client, including very large ones.

**Comment by oytuntez** [comment]
I am in a very similar situation with @EtienneAb3d, we cannot commit to v4 as it is now. Is there an expected date for license announcement?

@EtienneAb3d, retro legal changes are not possible, anything until license changes remains the same. I just forked the project to my personal account 🔐

_

_**speculation warning 😆**_
If you'll ask my more honest opinion, I think the license change may not be a ripe idea. I just started playing with MMT in my mind and started doing experiments. Again, all IMHO, I think a more holistic and historical approach would be to convince and support strong developers and software-heavy companies to contribute to MMT, positioning as THE open source Google Translate alternative for all use cases, not only for cases arising directly from translation industry.

**Comment by EtienneAb3d** [comment]
There are a lot of other open-source offers. I don't see a major reason that will make a company accept to pay for yours. It's hard to estimate from here, but I'm convinced that, from the moment you will have a payable license, you won't sell any one of it. 
The only real chance for such a project, is to gain the confidence of as much free users as possible, encouraging them to contribute, and then, sell something else like data processing, services, or even special dedicated proprietary devs.

**Comment by LoekvanKooten** [comment]
Actually I do see a reason for paid licenses: ModernMT is adaptive. No other tool is. Reading the above, they're after the big fish anyway, who currently pay thousands of euros for on-premise support. They'll need to come up with a really competitive offer for one-man shows like ours.

**Comment by EtienneAb3d** [comment]
Yes, MMT is the only available NMT tool offering this live-adaptive capability. The reason is certainly that no others are really seeing the real advantage/gain in doing so. They are rather trying to bet on general-to-specialised pre-adaptations. If this is true for specialists, it's also true for customers. It's very hard to convert a technical argument to a performant convincing commercial one! This is even more a problem with an install-it-yourself product!
I'm quite sure there are possibilities to sell installations/support/devs services, around the open-source free product. But, I'm very doubtful about the possibility to sell this product itself, in front of many other free solutions.

**Comment by LoekvanKooten** [comment]
Which once more confirms that the people buying these solutions simply have no clue about translation in general.

**Comment by oytuntez** [comment]
I believe ModernMT is one of the best solutions out there for mindsets like yours and ours – I definitely didn't mean otherwise, I find the paid license quite self-explanatory and fair. 

I just think there are better market and history positions for open source projects like this, with a broader perspective and impulse resistance 🥇

**Comment by mzeidhassan** [comment]
This is my personal opinion as one of the first users of MMT as far as I can tell.

I have been following this project since its inception. I remember Nicola's video on YouTube, even before there was source code available and I was impressed with how adaptive MMT is "the example about party in political context vs. generic". It was SMT back then.

While I understand why MMT is going this direction now, I think MMT has much better chance with small LSPs (Language Service Providers). The big players who have the capacity to hire developers, will have their own solution in place if they have to. They may also lean to some other open-source projects like OpenNMT or MarianMT. I admit that MMT in my opinion is the best because it offers a ready-to-use solution out of the box and it is adaptive. I have no doubt about this. It is really awesome. 

I think MMT team will have much better luck selling customized NMT solutions. For example, development of custom pre-/post-processing pipelines, MT plugins for MemoQ or Trados, selling bilingual data, custom-domain engines for different sectors, SAAS, etc. 

That being said, it may all depend on the licence fee and how much it will be. If it's reasonable, maybe some will be happy to cough up the money to use the tool. It's imperative for MMT though to make their solution open and free-to-use for private use and non-profit organizations. 

I know it's a hard decision to take, but I really wish you all the best, @davidecaroselli and @nicolabertoldi. You are very supportive and always willing to help. Thank you!

**Comment by cs16lover** [comment]
Hello @davidecaroselli ,

When will v4 release?

**Comment by davidecaroselli** [comment]
Hi all,

I have some important news. We finally took the time to look into the options we had and we have our final decision. **ModernMT v4.0 will be released totally open source, with Apache-2.0 license (the same of all the previous versions of our software)**.

**WHY?**
The major problem we had to face, was being able to contribute to open source and give our community the possibility to grow and create new tools with our technology as we have always done, but at the same time protect our work and create an asset in order to allow us to continue our work. At the end of the day, the solution is quite simple: we decided to publish all the new code, giving to our community full access and freedom of use, change and evolve of our machine translation tool.

**WHAT WILL CHANGE?**
As reported in our post, lots of new stuff!
- A brand new Neural Engine based on fairseq framework with **twice the translation speed**.
- A **boost in baseline models translation performance** measured in up to +2 BLEU points thanks to the new training pipeline.
- A brand **new aligner cleaning technique** to filter out noisy, un-aligned segments in training data for better performance.

**WHAT WILL BE THE DIFFERENCE WITH THE ENTERPRISE EDITION?**
What we offer in our Enterprise Edition is basically our inner knowledge of the tool: an **improved adaptation algorithm** "crafted" with months of optimization and fine-tuning of the system. Moreover, Enterprise Edition comes with **top-quality baseline models** trained on **billions of high-quality training data**.

These are all features built on top our our core technology that has always been, and will continue to be, totally **free and open source**!

**OK, COOL... BUT: WHEN WILL ALL THIS BE AVAILABLE?**
Very soon! Our Enterprise Edition software is complete and in production right now, we will shortly proceed migrating the code in this Open Source repository. We will then need to restore all the Command Line interface that needs to change to reflect the new software.
I hope everything to be ready in a month.

Thanks all for your feedbacks, and support!

**Comment by LoekvanKooten** [comment]
Sounds great, David! Thank you very much for everything you have done and are doing for this community.

**Comment by EtienneAb3d** [comment]
Great news !
Many thanks to all the MMT team !
😃

**Comment by oytuntez** [comment]
Wow, sounds great David! Cheers to the team.

**Comment by mzeidhassan** [comment]
Thank you so much @davidecaroselli for this great news. This is really amazing news. Thank you!

**Comment by davidecaroselli** [comment]
For the brave souls out there: there is a preview of the ModernMT v4 in the branch [versions/v4.0](https://github.com/modernmt/modernmt/tree/versions/v4.0).

If you want to help us spot bugs or any kind of problems before the release, please give it a try and report any problems in the issues page! Please, keep in mind that this is not ready for production and things could change in the final release.

Thanks for all your help!

**Comment by EtienneAb3d** [comment]
Found this error (easy to fix) when installing from sources:

```
[INFO] ------------------------------------------------------------------------
[INFO] Building Text Processing Module 4.0
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ mmt-textprocessing ---
[INFO] 
[INFO] --- build-helper-maven-plugin:1.12:cpu-count (get-cpu-count) @ mmt-textprocessing ---
[INFO] CPU count: 16
[INFO] 
[INFO] --- exec-maven-plugin:1.4.0:exec (generate-jflex) @ mmt-textprocessing ---
  File "/mnt/SSD/MMT4beta/modernmt-versions-v4.0/src/textprocessing/script/pyflex.py", line 188
    print 'Usage: pyflex.py SOURCE_DIRECTORY TARGET_DIRECTORY'
                                                             ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print('Usage: pyflex.py SOURCE_DIRECTORY TARGET_DIRECTORY')?
```

**Comment by nicolabertoldi** [comment]
thank you for spotting this error

Python3 requires a slightly different format of ``print`` instructions.

We will fix all of them as soon as possible.

**Comment by EtienneAb3d** [comment]
--gpus option no more available ?

**Comment by davidecaroselli** [comment]
Hi @EtienneAb3d 

we have just finished fixing both, can you please double check? Thanks!

**Comment by EtienneAb3d** [comment]
What kind of model ?  Convolutional ?

**Comment by davidecaroselli** [comment]
Just the regular one, no need to change the default:
```bash
./mmt create en it examples/data/train --gpus x y z
```

**Comment by EtienneAb3d** [comment]
My question was: of what kind is this default?

**Comment by davidecaroselli** [comment]
default is transformer, size "base"

**Comment by EtienneAb3d** [comment]
> we have just finished fixing both, can you please double check?

Both modifications seems to work (without the gpus option available, it first started on both available cards).

Is there a way to monitor in an other way than looking at the LOG file?

Is there a stop criterion?

**Comment by EtienneAb3d** [comment]
I provided with 30M sentence pairs. In the log, I read the GPU training is starting with 20M examples. 1/3 of my input data were removed?

**Comment by davidecaroselli** [comment]
@EtienneAb3d 
> Is there a way to monitor in an other way than looking at the LOG file?

Newer version of fairseq should provide tensorboard support. We will include it as soon as the new version is released

> Is there a stop criterion?

No, the decision is up to you by looking at the training perplexity

> I provided with 30M sentence pairs. In the log, I read the GPU training is starting with 20M examples. 1/3 of my input data were removed?

There are several steps of cleaning during the training , if you run the training with `--debug` you'll be able to look at all the temporary folders, one for every cleaning step.

Usually the major reduction comes from deduplication.

**Comment by EtienneAb3d** [comment]
Is there an option to make MMT4 keeping all the data I provide, without removing, nor deduplicating ?
Redundancy can be important to balance probabilities between similarities or rivalries.

**Comment by davidecaroselli** [comment]
if you run `./mmt create --help` you'll see option:

```
Data cleaning arguments:
  --skip-cleaning       skip the cleaning step (input corpora MUST be in plain
                        text parallel format)
```

**Comment by EtienneAb3d** [comment]
Installing from scratch on a new machine, I don't understand why "./mmt" is complaining about the lack of "fairseq", while it was installed by "setup.py", and even re-installed by hand with "requirements.txt" to be sure.
Any idea ?
```
Traceback (most recent call last):
  File "./mmt", line 75, in <module>
    main()
  File "./mmt", line 25, in main
    from cli import cleaning, datagen, train, create, server, translate, evaluate, memory
  File "/mnt/SSD/MMT4/modernmt/cli/datagen.py", line 14, in <module>
    from mmt.textencoder import SubwordDictionary
  File "/mnt/SSD/MMT4/modernmt/build/mmt-4.0.jar/mmt/__init__.py", line 4, in <module>
ModuleNotFoundError: No module named 'fairseq'
```

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

please, be sure to use pip3 and python3   (version 3.6 or higher) even during installation

please report the output of the following commands
```
python3 --version
pip3 --version
pip3 show fairseq torch
```

**Comment by EtienneAb3d** [comment]
```
python3 --version
Python 3.7.3

pip3 --version
pip 9.0.1 from /usr/lib/python3/dist-packages (python 3.6)

pip3 show fairseq torch
Name: fairseq
Version: 0.6.2
Summary: Facebook AI Research Sequence-to-Sequence Toolkit
Home-page: https://github.com/pytorch/fairseq
Author: UNKNOWN
Author-email: UNKNOWN
License: UNKNOWN
Location: /home/cubaix/.local/lib/python3.6/site-packages
Requires: tqdm, sacrebleu, cffi, numpy
---
Name: torch
Version: 1.0.1.post2
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: UNKNOWN
Author: UNKNOWN
Author-email: UNKNOWN
License: UNKNOWN
Location: /home/cubaix/.local/lib/python3.6/site-packages
Requires: 
```

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

there is a mismatch between the versions of python3 (3.7) and pip3 (3.6).

Basically during installation you installed fairseq for python3 (3.6), but then at run time you are using python3 (3.7).

You have two options:
- re-install ModernMT forcing the usage of pip3.7
- use python3 (version 3.6); you have probably to force python3 to point towards python3.6

They are equivalent, but most of our development was don using version 3.6 for both pip and python.
So I would suggest to do follow the second option

**Comment by EtienneAb3d** [comment]
I got it at the same time you replied ;-)
Doing this change on the first line of "./mmt" solve the problem:
`#!/usr/bin/env python3.6`

--steps option is no more recognized?

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 
your solution is also fine, but I would strongly suggest to have the same versions of pip3 and python3 to avoid further future misalignments.

I let @davidecaroselli answer about ``--steps``

**Comment by EtienneAb3d** [comment]
@nicolabertoldi 
I don't know how a got them both installed. Even, Synaptic is showing a "python3.7" package "not installed"!
I made a lot of attempts, without success for the moment, to make old MMT2 and MMT3 running on this new machine. What was done is a bit confuse now.
:)

**Comment by LoekvanKooten** [comment]
Is retraining needed for 4.0, or can memories from the previous 3.x version be used too?

**Comment by davidecaroselli** [comment]
@LoekvanKooten ModernMT v4.0 will use a totally different neural network framework (from Tensorflow to Fairseq) so yes, a re-training is needed.

**Comment by EtienneAb3d** [comment]
There is certainly a problem with the embedding of the text parameter in the `./mmt translate` command: when a text start with a '-', MMT complains about a bad option, and doesn't like having spaces in the text.

It seems MMT4 doesn't like having language codes with countries, like "fr-FR" or "en-GB". It's a problem a also got with MMT3. Is it possible to do the same fix on MMT4, like on MMT3, to be able to have country codes?

**Comment by davidecaroselli** [comment]
> There is certainly a problem with the embedding of the text parameter in the ./mmt translate

This is not a problem with `translate` command, this is how the Bash shell works. If you need to specify a **single** parameter that has spaces in it, you have to use double quotes, like this:
```bash
$ ./mmt translate hello world
usage: mmt translate [-h] [-s SOURCE_LANGUAGE] [-t TARGET_LANGUAGE]
                     [--context CONTEXT] [--context-file CONTEXT_FILE]
                     [--context-vector CONTEXT_VECTOR] [-e ENGINE] [--batch]
                     [--xliff]
                     [TEXT]
mmt translate: error: unrecognized arguments: world

$ ./mmt translate "hello world"
ciao mondo
```

While if you want to include special chars (like "-" that in bash shell standard means "argument start"), you have to pass it in stdin. from a `echo` command or a file to translate:
```bash
$ echo "-hello world" | ./mmt translate --batch
-ciao mondo

$ echo "-hello world" > file_to_translate.txt
$ echo "this is an example" >> file_to_translate.txt
$ ./mmt translate --batch < file_to_translate.txt
-ciao mondo
questo è un esempio
```

> It seems MMT4 doesn't like having language codes with countries

Not sure what you mean with this, what problem did you find? for example I can translate using these languages:
```bash
$ ./mmt translate -s en-US -t it-IT "hello world"
ciao mondo
```

And I can of course specify countries in the `model.conf`:
```bash
$ cat engines/default/models/decoder/model.conf
[models]
en__it-IT = en__it/

$ cat engines/default/engine.xconf 
<node xsi:schemaLocation="http://www.modernmt.eu/schema/config mmt-config-1.0.xsd"
      xmlns="http://www.modernmt.eu/schema/config"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <engine source-language="en" target-language="it-IT" />
</node>

$ ./mmt translate -s en-US -t it-IT "hello world"
ciao mondo
```

**Comment by EtienneAb3d** [comment]
The country code problem should be in the memory part of code. Without a context, it works, because the memory part is not involved (isn't it ?). Try the same command with a context.

**Comment by EtienneAb3d** [comment]
Fairseq is very RAM consuming: about 0.5G for each 1M sentence pair. I tried a 250M pairs training: 120G in RAM, thus, with the system, it's just starting swapping with my 128G available. This certainly will make a bit hard to use the machine for something else during several days of training. :(

**Comment by LoekvanKooten** [comment]
How do I solve this?

`loek@DORAEMON:~/mmt9/modernmt$ python3.6 setup.py
torch-1.0.1.post2-cp36-cp36m-linux_x86_64.whl is not a supported wheel on this platform.
You are using pip version 8.1.1, however version 19.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Traceback (most recent call last):
  File "setup.py", line 248, in <module>
    main()
  File "setup.py", line 236, in main
    pip_install()
  File "setup.py", line 219, in pip_install
    osutils.shell_exec(['pip3', 'install', '-r', requirements_txt], stderr=sys.stderr, stdout=sys.stdout)
  File "/home/loek/mmt9/modernmt/cli/utils/osutils.py", line 60, in shell_exec
    raise ShellError(str_cmd, return_code, stderr_dump)
cli.utils.osutils.ShellError: Command 'pip3 install -r /home/loek/mmt9/modernmt/requirements.txt' failed with exit code 1
`

**Comment by nicolabertoldi** [comment]
@LoekvanKooten 

please try to edit the file ``requirements.txt``
in this way
```
regex
sacrebleu==1.3.1
requests
cachetools
torch
torchvision
fairseq==0.6.2
```

and then rerun the command
```
python3.6 setup.py
```

and let us know if you succeed

**Comment by davidecaroselli** [comment]
Guys please, open different issues for different problems, or this is going to be a mess in a couple of posts :)

**Comment by LoekvanKooten** [comment]
That worked! New issue though. Will open new thread.

**Comment by davidecaroselli** [comment]
@EtienneAb3d I found some problems more generally with context vector parsing, I will write you back asap.

**Comment by EtienneAb3d** [comment]
Remark: on my side, I got a problem with the context vector, but this was due to the use of the French number formatting using a "," in place of "." for decimal on this new machine, thus not a problem coming from MMT4.

**Comment by davidecaroselli** [comment]
@EtienneAb3d just pushed a fix for buggy ContextVector parsing. If you can please check and report any language-tag related problems it would be great! (Just, open a different issue if the case, so we do not mess up this one)

Thanks!

**Comment by davidecaroselli** [comment]
ModernMT v4.0 is finally available!
Thanks all for the patience and your support!
